base_layer/core/src/validation/header/header_full_validator.rs

```diff
@@ -89,20 +89,24 @@ impl<B: BlockchainBackend> HeaderChainLinkedValidator<B> for HeaderFullValidator
     }
 }
 
-fn check_timestamp_count(
+/// This is a sanity check for the information provided by the caller, rather than a validation for the header itself.
+fn sanity_check_timestamp_count(
     header: &BlockHeader,
-    prev_timestamps: &[EpochTime],
+    timestamps: &[EpochTime],
     consensus_constants: &ConsensusConstants,
 ) -> Result<(), ValidationError> {
-    let expected_timestamp_count = cmp::min(
-        consensus_constants.median_timestamp_count(),
-        usize::try_from(header.height - 1)
-            .map_err(|_| ValidationError::CustomError("Invalid conversion u64 to uszie".to_string()))?,
-    );
-    let timestamps: Vec<EpochTime> = prev_timestamps.iter().take(expected_timestamp_count).copied().collect();
-    if timestamps.len() < expected_timestamp_count {
-        return Err(ValidationError::NotEnoughTimestamps {
-            actual: timestamps.len(),
+    let expected_timestamp_count = cmp::min(consensus_constants.median_timestamp_count() as u64, header.height);
+    // Empty `timestamps` is never valid
+    if timestamps.is_empty() {
+        return Err(ValidationError::IncorrectNumberOfTimestampsProvided {
+            expected: expected_timestamp_count,
+            actual: 0,
+        });
+    }
+
+    if timestamps.len() as u64 != expected_timestamp_count {
+        return Err(ValidationError::IncorrectNumberOfTimestampsProvided {
+            actual: timestamps.len() as u64,
```

Comment:
Yup the blockchain db code is super hard to reason about. It's not strictly a header validation since the timestamps come from headers previously inserted into the main/orphan chains so it should be correct because our code produced it. This is why I renamed it to sanity_check because it's more about asserting an invariant we require when calling the validator. The other option is to take it out completely or to panic, but I went with the least risky approach.